geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/filters/Z3Filter.scala

```diff
@@ -60,20 +123,87 @@ class Z3Filter(
     }
   }
 
+  private def nextJumpIn(epoch: Short, z: Long): FilterResult = {
+    val epochIndex = epoch - epochOffset
+    val zb = if (epochIndex < 0 || epochIndex >= zBounds.length) { zBoundsNoTime } else { zBounds(epochIndex) }
+
+    var nextZ = Long.MaxValue
+
+    var i = 0
+    while (i < zb.bounds.length) {
+      val Array(zmin, zmax) = zb.bounds(i)
+      if (z < zmin) {
+        if (zmin < nextZ) {
+          nextZ = zmin
+        }
+      } else if (z < zmax) {
+        val next = Z3.zdivide(z, zmin, zmax)._2
+        if (next < nextZ) {
+          nextZ = next
+        }
+      }
+      i += 1
+    }
+
+    var nextEpoch = epoch
+
+    if (nextZ == Long.MaxValue && epoch != Short.MaxValue) {
+      val nextEpochIndex = epochIndex + 1
+      if (nextEpochIndex < 0 || nextEpochIndex >= zBounds.length) {
+        nextZ = zBoundsNoTime.min
+      } else {
+        nextZ = zBounds(nextEpochIndex).min
+      }
+      nextEpoch = (epoch + 1).toShort
+    }
+
+    logger.trace(s"Seeking ahead from $epoch:$z to $nextEpoch:$nextZ")
+    FilterResult.SkipAhead(ByteArrays.toBytes(nextEpoch, nextZ))
+  }
+
   override def toString: String = Z3Filter.serializeToStrings(this).toSeq.sortBy(_._1).mkString(",")
 }
 
 object Z3Filter extends RowFilterFactory[Z3Filter] {
 
+  import org.locationtech.geomesa.index.conf.QueryHints.RichHints
+
   private val RangeSeparator = ":"
   private val TermSeparator  = ";"
   private val EpochSeparator = ","
 
-  val XYKey     = "zxy"
-  val TKey      = "zt"
-  val EpochKey  = "epoch"
+  val VersionKey = "v"
+  val XYKey      = "zxy"
+  val TKey       = "zt"
+  val EpochKey   = "epoch"
+  val MaxTimeKey = "mt"
+  val SeekKey    = "s"
+
 
-  def apply(values: Z3IndexValues): Z3Filter = {
+  // note: we use a negative version number to ensure that we error out with a negative length if an older version
+  // of the filter tries to deserialize a newer serialized buffer. This was verified with the code snippet below
```

Comment:
Ok, so what does the error look like?

Also, sounds like this is not backwards compatible, right?

Reply:
it would be some kind of error from running `Array.ofDim(-1)`. so not exactly helpful, but it would ensure that things fail fast.

it is backwards compatible in that the newer iterator should work with older clients, but it is not backwards compatible in that older iterators work with newer clients. we don't have a good way to detect the iterator version in order to send the older serialization format...